linter/lints/common.go

```diff
@@ -35,3 +37,50 @@ func GetExtWithOID(exts []pkix.Extension, oid asn1.ObjectIdentifier) *pkix.Exten
 	}
 	return nil
 }
+
+// ReadOptionalASN1BooleanWithTag attempts to read and advance incoming to
+// search for an optional DER-encoded ASN.1 element tagged with the given tag.
+// Unless outExpectedTagPresent or outPresent are nil, it stores whether the tag
+// was found in outExpectedTagPresent and an element with the tag was found in
+// outPresent, otherwise each boolean will take the default value. It reports
+// whether all reads were successful.
+func ReadOptionalASN1BooleanWithTag(incoming *cryptobyte.String, outExpectedTagPresent *bool, outPresent *bool, tag cryptobyte_asn1.Tag, defaultValue bool) bool {
+	if outExpectedTagPresent != nil {
+		*outExpectedTagPresent = defaultValue
+	}
+	if outPresent != nil {
+		*outPresent = defaultValue
```

Comment:
As a matter of nice idiom, a function that errors shouldn't have side effects (when possible). For instance, if we fail to parse the ASN.1 input, we shouldn't have a side effect of writing to the outparam with the default value. Instead, we should have a _local_ variable that contains the default value and is overwritten with the actually-read value if appropriate. Then before returning `true`, we write to `*out`.

Reading [the source of `ReadOptionalASN1`](https://cs.opensource.google/go/x/crypto/+/refs/tags/v0.13.0:cryptobyte/asn1.go;l=658), I see that it doesn't follow this nice idiom. It writes to `*outPresent` regardless of whether the subsequent `ReadASN1` succeeds. Not sure why! Perhaps it's because there is an internal guarantee: if `PeekASN1Tag` succeeds, then `ReadASN1` is guaranteed to succeed. I don't know if that's an internal guarantee or what. A bit of a mystery.

Reply:
I relied heavily on `ReadOptionalASN1` and `ReadOptionalASN1OctetString` while writing this and styled my function off of those. I now really like out parameters 🤷🏼. I wanted the two bool outparams so that I could better distinguish what part of `ReadOptionalASN1BooleanWithTag` failed during linting.